sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1744 @@
+    var diskHandle: DiskHandle = null
+    synchronized {
+      if (host.isDefined) {
+        hostBuffer = host.get
+        hostBuffer.incRefCount()
+      } else if (disk.isDefined) {
+        diskHandle = disk.get
+      } else {
+        throw new IllegalStateException(
+          "attempting to materialize a closed handle")
+      }
+    }
+    if (hostBuffer != null) {
+      GpuTaskMetrics.get.readSpillFromHostTime {
+        withResource(hostBuffer) { _ =>
+          dmb.copyFromHostBuffer(
+            /*dstOffset*/ 0,
+            /*src*/ hostBuffer,
+            /*srcOffset*/ 0,
+            /*length*/ hostBuffer.getLength)
+        }
+      }
+    } else {
+      // cannot find a full host buffer, get chunked api
+      // from disk
+      diskHandle.materializeToDeviceMemoryBuffer(dmb)
+    }
+  }
+
+  private[spill] def setHost(singleShotBuffer: HostMemoryBuffer): Unit = synchronized {
+    host = Some(singleShotBuffer)
+  }
+
+  private[spill] def setDisk(handle: DiskHandle): Unit = synchronized {
+    disk = Some(handle)
+  }
+}
+
+object SpillableDeviceBufferHandle {
+  def apply(dmb: DeviceMemoryBuffer): SpillableDeviceBufferHandle = {
+    val handle = new SpillableDeviceBufferHandle(dmb.getLength, dev = Some(dmb))
+    SpillFramework.stores.deviceStore.track(handle)
+    handle
+  }
+}
+
+class SpillableDeviceBufferHandle private (
+    val sizeInBytes: Long,
+    private[spill] override var dev: Option[DeviceMemoryBuffer],
+    private[spill] var host: Option[SpillableHostBufferHandle] = None)
+    extends DeviceSpillableHandle[DeviceMemoryBuffer] {
+
+  override val approxSizeInBytes: Long = sizeInBytes
+
+  private[spill] override def spillable: Boolean = synchronized {
+    if (super.spillable) {
+      dev.getOrElse {
+        throw new IllegalStateException(
+          s"$this is spillable but it doesn't have a dev buffer!")
+      }.getRefCount == 1
+    } else {
+      false
+    }
+  }
+
+  def materialize(): DeviceMemoryBuffer = {
+    var materialized: DeviceMemoryBuffer = null
+    var hostHandle: SpillableHostBufferHandle = null
+    synchronized {
+      if (host.isDefined) {
+        // since we spilled, host must be set.
+        hostHandle = host.get
+      } else if (dev.isDefined) {
+        materialized = dev.get
+        materialized.incRefCount()
+      } else {
+        throw new IllegalStateException(
+          "attempting to materialize a closed handle")
+      }
+    }
+    // if `materialized` is null, we spilled. This is a terminal
+    // state, as we are not allowing unspill, and we don't need
+    // to hold locks while we copy back from here.
+    if (materialized == null) {
+      materialized = closeOnExcept(DeviceMemoryBuffer.allocate(sizeInBytes)) { dmb =>
+        hostHandle.materializeToDeviceMemoryBuffer(dmb)
+        dmb
+      }
+    }
+    materialized
+  }
+
+  override def spill(): Long = {
+    if (!spillable) {
+      0L
+    } else {
+      synchronized {
```

Comment:
Is there a race condition here? If Thread t1 calls `spill()`, and they see that `spillable` is true, but then thread t2 jumps in and calls `materialize()` and gets the lock and materializes. Say the buffer is on the device only at the moment, so t2 gets a ref to the `DeviceMemoryBuffer`, and ref count goes to 2. Now t2 exits, and t1 proceeds and starts spilling even though it's no longer spillable.

Reply:
Yes, this is an acceptable race.

In this case we copy to host but we can't free (we call close but we don't actually free). From the perspective of the spill framework, we are going to synchronize up to the point where we call .close on the `DeviceMemoryBuffer`. The rest is  left to the caller. The caller needs to synchronize w.r.t. apis that it calls, and I think that part is understood (it's part of the original spill framework as well.